design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,193 @@
+- Whenever the `/etc/tedge/config-plugins` directory is updated (new plugin installed or existing one removed),
+  agent refreshes the supported types.
+  Simply touching this directory would also trigger a refresh.
+- When the supported types are published over mqtt, their source plugin information is also appended to that type
+  in the format <config_type>::<plugin_type>.
+  For example, if a `mosquitto` plugin lists two different config types: `mosquitto.conf` and `mosquitto.acl`,
+  both types would be reported as `mosquitto.conf::mosquitto` and `mosquitto.acl::mosquitto`.
+- When a `config_snapshot` request is received for a type, call the `get` command of the correspinding plugin for that type
+  which can be derived from the `::` suffix of that type.
+  If there is no explicit suffix, it is delegated to the default file plugin.
+- When a `config_update` request is received for a type, the following actions are performed in sequence:
+  - Call `get` command of the corresponding plugin and cache the target temporary file as a backup.
+  - Call `set` command of the corresponding plugin with the updated config path in the argument.
+  - Call `finalize` command of the plugin and complete the operation if successful.
+  - If `set` or `finalize` commands fail, call the `set` command again with the original config file backup as the target.
+  - Call `finalize` again if the previous `set` attempt to restore teh original configuration succeeded and fail the operation.
+
+### Phase 2
+
+Instead of the `exeucting` phase of `config_update` workflow doing everything in that single step,
+break it down into multiple smaller stages that corresponds to each plugin command/phase as well as follows:
+
+```config_update.toml
+operation = "config_update"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+[scheduled]
+action = "proceed"
+on_success = "executing"
+
+[executing]
+action = "builtin:config_update:executing"
+on_success = "download"
+
+[download]
+action = "builtin:download"
+on_success = "prepare"
+
+[prepare]
+action = "builtin:backup_file"
+on_success = "apply"
+
+[apply]
+action = "builtin:set_file"
+on_success = "validate"
+
+[validate]
+action = "builtin:config_update:validate"
+on_success = "commit"
+on_error = "rollback"
+
+[commit]
+action = "builtin:config_update:apply"
+on_success = "successful"
+
+[rollback]
+action = "builtin:config_update:rollback"
+on_success = "failed"
+
+[successful]
+action = "cleanup"
+
+[failed]
+action = "cleanup"
+```
+
+This helps customers override a single stage like `download` to do it their way,
+while still reusing the rest of the functionality as-is.
+
+### Unresolved problems
+
+1. When a new piece of software is installed on the device,
+   which mostly would have come with its own configurations and log files,
+   how to "notify tedge" so that it refreshes its
+   - updated software list
+   - supported config list
+   - updated logs list
```

Comment:
If I read it correctly, 
1. updated software list -> we trigger `software_list` operation after `software_update` operation is done. So I think no problem.
2. & 3. supported config/log lists -> we can trigger such operations (or "signals"), requesting to resend supported config/log types after `software_update` operations. It might be related to my current MQTT API work.

Reply:
Yes. With the proposed "signals", I'm even inclined to make the agent declare its `software_list` with a `sig` instead of an explicit `cmd`. Using a `cmd` for `software_list` with its state transitions was unnecessary right from the start. A `sig` would have been the better choice, if it existed back then.